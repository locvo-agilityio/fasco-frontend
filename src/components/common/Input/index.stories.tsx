import type { Meta, StoryObj } from '@storybook/react';

// Components
import Input from '.';

const meta = {
  title: 'Components/Common/Input',
  component: Input,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Input>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    placeholder: 'John Carter',
  },
};

export const HasError: Story = {
  args: {
    isInvalid: true,
    errorMessage: 'Error message',
    placeholder: 'John Carter',
    defaultValue: 'Error Data',
  },
};
