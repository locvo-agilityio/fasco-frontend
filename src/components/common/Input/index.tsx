import { Input as InputNextUI } from '@heroui/react';
import { ChangeEvent, ComponentProps, memo } from 'react';

interface TInputProps
  extends Omit<ComponentProps<typeof InputNextUI>, 'onChange'> {
  errorMessage?: string;
  onChange?: (value: string) => void;
}

const Input = ({
  isInvalid = false,
  errorMessage = 'Default error',
  onChange,
  ...rest
}: TInputProps) => {
  const inputStyle = [
    'group-data-[has-value=true]:text-gray-500',
    'placeholder:text-gray-500',
    'text-gray-500',
    'text-base',
  ];
  const inputWrapperStyle = ['after:bg-transparent'];
  const labelStyle = ['text-gray-500', 'text-base'];

  const handleChangeValue = (e: ChangeEvent<HTMLInputElement>) => {
    onChange?.(e.target.value);
  };

  return (
    <div className="flex w-full flex-wrap md:flex-nowrap mb-6 md:mb-0 gap-4">
      <InputNextUI
        variant="underlined"
        classNames={{
          input: inputStyle,
          inputWrapper: inputWrapperStyle,
          label: labelStyle,
        }}
        onChange={handleChangeValue}
        isInvalid={isInvalid}
        errorMessage={errorMessage}
        {...rest}
      />
    </div>
  );
};

export default memo(Input);
