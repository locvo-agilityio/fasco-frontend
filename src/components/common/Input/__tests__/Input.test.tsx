import { fireEvent, render } from '@testing-library/react';

// Components
import Input from '..';

describe('Input component', () => {
  const mockOnChange = jest.fn();
  const props = {
    label: 'Email',
    name: 'email',
    placeholder: 'Email',
    onChange: mockOnChange,
  };
  it('render correctly with default props', () => {
    const { container } = render(<Input {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('render correctly with error message', () => {
    const { getByText } = render(
      <Input {...props} isInvalid errorMessage="Field is required" />,
    );
    expect(getByText('Field is required')).toBeInTheDocument();
  });

  it('should call onChange when input value is changed', async () => {
    const { getByPlaceholderText } = render(<Input {...props} />);
    fireEvent.change(getByPlaceholderText('Email'), {
      target: { value: 'user@gmail.com' },
    });

    expect(mockOnChange).toHaveBeenCalled();
  });
});
