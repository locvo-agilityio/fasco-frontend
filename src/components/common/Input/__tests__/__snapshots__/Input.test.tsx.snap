// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component render correctly with default props 1`] = `
<div>
  <div
    class="flex w-full flex-wrap md:flex-nowrap mb-6 md:mb-0 gap-4"
  >
    <div
      class="group flex flex-col data-[hidden=true]:hidden w-full"
      data-filled="true"
      data-filled-within="true"
      data-has-elements="true"
      data-has-helper="true"
      data-has-label="true"
      data-slot="base"
    >
      <div
        class="w-full inline-flex tap-highlight-transparent px-3 !px-1 !pb-0 !gap-0 relative box-border border-b-medium shadow-[0_1px_0px_0_rgba(0,0,0,0.05)] border-default-200 !rounded-none hover:border-default-300 after:content-[''] after:w-0 after:origin-center after:absolute after:left-1/2 after:-translate-x-1/2 after:-bottom-[2px] after:h-[2px] group-data-[focus=true]:after:w-full min-h-10 rounded-medium flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 after:transition-width motion-reduce:after:transition-none h-14 py-2 after:bg-transparent"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <label
          class="absolute z-10 pointer-events-none origin-top-left flex-shrink-0 rtl:origin-top-right subpixel-antialiased block cursor-text will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.small)/2_-_3.5px)] pe-2 max-w-full text-ellipsis overflow-hidden text-gray-500 text-base"
          data-slot="label"
          for="react-aria-:r0:"
          id="react-aria-:r1:"
        >
          Email
        </label>
        <div
          class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-1.5"
          data-slot="inner-wrapper"
        >
          <input
            aria-label="Email"
            aria-labelledby="react-aria-:r0: react-aria-:r1:"
            class="w-full font-normal bg-transparent !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text group-data-[has-value=true]:text-gray-500 placeholder:text-gray-500 text-gray-500 text-base"
            data-slot="input"
            id="react-aria-:r0:"
            name="email"
            placeholder="Email"
            title=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
